### DISCUSSION TITLE
How to create a sensor for alerting on asset check runs?

### DISCUSSION THREAD

**Discussion by dennisgera** [dagster-io/dagster#19944] [Q&A] [answered]
Hi team,

I currently use OS Dagster and want to build a sensor that can trigger a job (send slack message) whenever an asset check fails. How would I implement that?

**Comment by garethbrickman** [comment]
There is a [dagster-slack library](https://docs.dagster.io/_apidocs/libraries/dagster-slack) and within it there is a [make_slack_on_run_failure_sensor](https://docs.dagster.io/_apidocs/libraries/dagster-slack#dagster_slack.make_slack_on_run_failure_sensor) function that creates a sensor on job failures that will message the given Slack channel.

**Reply by dennisgera** [reply]
This is for job failures. I'm looking for asset check failures.

**Reply by garethbrickman** [reply]
Ah, there doesn't seem to be an out-of-the-box solution for asset check failures in Dagster Open Source, so it'd have to be hand-rolled. 

Asset checks are included in [Dagster Cloud's Alerts feature](https://docs.dagster.io/dagster-cloud/managing-deployments/setting-up-alerts#setting-up-alerts-in-dagster-cloud).

**Reply by dennisgera** [reply]
Any idea how I'd start this hand-rolled solution?

**Reply by dennisgera** [reply]
@garethbrickman , I ended up with a solution as such:

```python

def make_asset_check_failure_sensor(
    name: str,
    job,
    asset_check_names: list,
    default_status: DefaultSensorStatus,
    run_config=None,
    run_config_fn=None,
):
    @sensor(
        name=name,
        job=job,
        default_status=default_status,
    )
    def _asset_check_failure_sensor(context: SensorEvaluationContext):
        instance: DagsterInstance = context.instance
        last_cursor = context.cursor.split("_")[0] if context.cursor else "0"
        after_timestamp = datetime.now() - timedelta(hours=3)

        filters = EventRecordsFilter(
            event_type=DagsterEventType.ASSET_CHECK_EVALUATION,
            after_cursor=RunShardedEventsCursor(id=int(last_cursor), run_updated_after=after_timestamp),
        )
        event_records = instance.get_event_records(event_records_filter=filters)

        for event_record in event_records:
            if is_asset_check_failure(event_entry=event_record.event_log_entry, asset_check_names=asset_check_names):
                _run_config = run_config_fn(context, event_record) if run_config_fn else run_config
                key = f"{event_record.storage_id}_{event_record.run_id}"

                context.update_cursor(key)
                yield RunRequest(run_key=key, run_config=_run_config)

    return _asset_check_failure_sensor
```

Please comment if you find this solution adequate, given how Dagster manipulates its event logs. However, I find it weird that I'm getting the same `storage_id`s on multiple occasions for different asset check runs with the same asset check. Not sure if that's expected when testing dagster locally.

**Reply by dennisgera** [reply]
For some reason, my cursor also wouldn't update on every new run - which I could see had a different `run_key` than on the last cursor.

**Comment by josh-gree** [comment]
I have a selection of assets - do not explicitly use jobs at all - I would like to be able to send messages on asset materialisations - how is this accomplished? Seems to only be possible on a job?

**Reply by pcadmanbosse** [reply]
@josh-gree You could use an asset sensor perhaps?

**Comment by NicolasGuary** [comment, marked as answer]
Hey ! I posted here a solution I implemented based on some replies here : https://github.com/dagster-io/dagster/discussions/21281
Hope this helps, I'm using it for a few weeks now without issues!

**Reply by garethbrickman** [reply]
That's brilliant, well done and thank you!
